### PR TITLE
[Perf] Tiles 4c: add Tiles16x16 proxy

### PR DESCRIPTION
**Pull request by hughperkins**
Issue: #

## Summary
  Introduces the public `qd.simt.Tile16x16` proxy API and migrates all tests to use it, replacing the private `_make_tile16x16()` factory. Also simplifies kernel signatures by
  reverting `tile_size` from a kernel parameter back to a local variable.
  ### What's in this PR
  **New public API (`_Tile16x16Proxy`)**
  - `qd.simt.Tile16x16.zeros(dtype=qd.f32)` / `qd.simt.Tile16x16.eye(dtype=qd.f64)` — create tiles with explicit dtype
  - `qd.simt.Tile16x16.zeros()` — omit dtype to use the runtime's `default_fp`
  - `qd.simt.Tile16x16.SIZE` — tile dimension constant (16)
  - Lazy-loaded via `__getattr__` in `simt/__init__.py` to avoid circular imports
  - CPU backend detection: raises `QuadrantsSyntaxError` with a clear message
  **`tile_size` reverted from kernel parameter to local variable**
  - In tiles-4b, `tile_size: qd.Template` was added as an explicit kernel parameter (passed as `_TILE` at call sites) to avoid accessing `qd.simt.Tile16x16.SIZE` inside
  `fastcache=True` kernels
  - This PR reverts that: all kernels now use `tile_size = qd.simt.Tile16x16.SIZE` as a local variable, and `tile_size` is removed from kernel signatures and call sites
  - This simplifies every kernel signature and call site, and is validated by every test in the file passing
  **Test migration to public API**
  - All ~32 existing kernel tests migrated from `_make_tile16x16()`/`Tile()` to `qd.simt.Tile16x16.zeros()`/`eye()`
  - `_make_tile16x16` import retained only for `test_tile16_make_caching` and `test_tile16_solve_triangular_upper_raises`
  **17 new tests for the proxy API**
  - `test_tile16_load_store_partial`, `test_tile16_load_store_partial_rows` — partial column/row load-store
  - `test_tile16_syr_sub` — symmetric rank-1 subtract via scalar elements
  - `test_tile16_potrf`, `test_tile16_potrf_then_trsm` — Cholesky and combined Cholesky+trsm using `eps_field` and `_EPS_VALS`
  - `test_tile16_vec_proxy_syr_sub_2d/3d`, `test_tile16_vec_proxy_partial_rows`, `test_tile16_vec_proxy_multi_column_accumulate` — vec proxy patterns with `tensor_type`
  parametrization
  - `test_tile16_f64_roundtrip_into_f32_array` — f64 tile on f32 data
  - `test_tile16_raises_on_cpu` — CPU backend error
  - `test_proxy_size_constant`, `test_simt_invalid_attr_raises` — basic proxy attributes
  - `test_proxy_default_dtype`, `test_proxy_eye_explicit_dtype` — dtype resolution
  - `test_proxy_in_func` — proxy inside `@qd.func`
  - `test_proxy_default_dtype_survives_reinit` — dtype correctness across `qd.reset()` cycles
  **Docs updated** to use `qd.simt.Tile16x16` throughout (quick start, creating tiles, identity init, kernel structure, f64 support, method reference table). Added combined
  Cholesky+trsm example.
  **Other**
  - `is_from_quadrants_module` in `util.py`: moved `import types` to top-level
  - `_make_tile16x16` docstring updated to recommend using the proxy
  - Type stub (`_Tile16x16Proto`) added for pyright/IDE support
  ### Good
  - Users no longer need to import a private `_make_tile16x16` factory — `qd.simt.Tile16x16` is discoverable and documented
  - Kernel signatures are simpler: no more `tile_size: qd.Template` parameter or `_TILE` at call sites
  - `qd.simt.Tile16x16.SIZE` is accessed inside every kernel in the test suite (via `tile_size = qd.simt.Tile16x16.SIZE`), so the pattern is thoroughly exercised
  - CPU backend gives a clear error message instead of a crash
  - Default dtype support (`zeros()` without args) follows `default_fp`
  - Extensive test coverage including edge cases like `reset()`/`reinit()` cycles
  - Docs fully updated to match the new API
  ### Bad
  - `qd.simt.Tile16x16` is a module-level singleton instance (not a class), so the purity checker's `is_from_quadrants_module` returns `False` for it — accessing
  `qd.simt.Tile16x16.SIZE` inside a `fastcache=True` kernel emits a non-fatal `PURE.VIOLATION` warning. This doesn't affect correctness or CI (warnings are not treated as
  errors), but is not fully clean. A future fix to `is_from_quadrants_module` to recognize proxy instances would eliminate this
  - The `_Tile16x16Proto` type stub duplicates method signatures and will need manual upkeep if the tile API changes
  - `_make_tile16x16` and its cache are still the underlying mechanism — the proxy is a thin wrapper, not a replacement of the internals

copilot:summary

### Walkthrough

copilot:walkthrough

### CODE REVIEW

File: tests/python/test_tile16.py [hunk 1]
```diff
@@ -1097,3 +1097,513 @@ def k1(s: qd.types.NDArray[qd.f32, 2], d: qd.types.NDArray[qd.f32, 2], tile_size
...
+@pytest.mark.parametrize("tensor_type", [qd.ndarray, qd.field])
+@pytest.mark.parametrize("qd_dtype", _QD_DTYPES)
+@test_utils.test(arch=qd.gpu)
+def test_tile16_syr_sub(tensor_type, qd_dtype):
+    test_utils.skip_if_f64_unsupported(qd_dtype)
+    np_dtype = _NP_DTYPES[qd_dtype]
+    mat = tensor_type(qd_dtype, (_TILE, _TILE))
+    vec = tensor_type(qd_dtype, (_TILE,))
+    out = tensor_type(qd_dtype, (_TILE, _TILE))
+
```
[the remark this answers is not in the document]
Comment (hughperkins) [reply tests/python/test_tile16.py:1136]: from Opus:

  The reviewer is right on point (2) but wrong on point (1).

  Point (1) — for tid in range(tile_size) is NOT a bug. The reviewer is misunderstanding Quadrants' kernel model. In Quadrants/Taichi, the top-level for loop IS the
  parallelization primitive — each iteration runs on a different thread. With block_dim=16 and for tid in range(16), each of the 16 threads gets its own tid value (0–15). It
  does NOT mean all 16 threads share the same loop counter. So vec_arr[tid] gives each thread a different scalar, and qd.outer(vec_arr[tid], vec_arr[tid]) correctly computes the
   rank-1 symmetric outer product.

  You can verify this by looking at test_tile16_slice_ger_sub_via_outer, which does essentially the same thing — qd.outer(per_thread_scalar, per_thread_scalar) — and correctly
  produces a rank-1 outer product.

  That said, using for _ in range(tile_size) + tid = qd.simt.subgroup.invocation_id() would be more consistent with the convention used in every other test in the file. So the
  suggested change is fine as a style improvement, just not a correctness fix.

  Point (2) — atol=_ATOLS[qd.f32] is a real bug. The test is parametrized over qd_dtype (both f32 and f64), but the assertion hardcodes the f32 tolerance. Should be
  atol=_ATOLS[qd_dtype]. This makes the f64 path 100,000x too lenient (1e-5 instead of 1e-10), so it would pass but wouldn't catch regressions.

  I'd fix point (2) and optionally apply point (1) for stylistic consistency. Want me to make the changes?

[the remark this answers is not in the document]
Comment (hughperkins) [reply tests/python/test_tile16.py:1136]: Opus:

  Done. Pushed commit 92ec11890 to hp/tiles-4c — changed _ATOLS[qd.f32] to _ATOLS[qd_dtype] in test_tile16_syr_sub.

**Review by erizmr — APPROVED** [review]
Approving on the basis that I have reviewed the design and the public facing API, tests, and they look reasonable to me.

**Comment by hughperkins** [comment]
Thanks!
